cli/azd/pkg/environment/environment.go

```diff
@@ -144,13 +139,38 @@ func (e *Environment) Getenv(key string) string {
 	return os.Getenv(key)
 }
 
+// Reloads environment variables from .env file
+func (e *Environment) Reload() error {
+	envPath := filepath.Join(e.Root, azdcontext.DotEnvFileName)
+	if envMap, err := godotenv.Read(envPath); errors.Is(err, os.ErrNotExist) {
+		e.Values = make(map[string]string)
+	} else if err != nil {
+		return fmt.Errorf("loading .env: %w", err)
+	} else {
+		e.Values = envMap
+	}
+
+	return nil
+}
+
 // If `Root` is set, Save writes the current contents of the environment to
 // the given directory, creating it and any intermediate directories as needed.
 func (e *Environment) Save() error {
 	if e.Root == "" {
 		return nil
 	}
 
+	// Cache current values & reload to get any new env vars
+	currentValues := e.Values
+	if err := e.Reload(); err != nil {
+		return fmt.Errorf("failed reloading env vars, %w", err)
+	}
+
+	// Overlay current values before saving
+	for key, value := range currentValues {
+		e.Values[key] = value
+	}
```

Comment:
what if there is a new value after `Reload` for a key in the `currentValues` ?  Are we basically overriding any new value with current? 
Should the current values be overlay with the current values instead?

Reply:
We cache the current state of the env object before reloading.  The reload will add any new env vars that were not in the original env state and then we overlay the current env state back before the Save() operation.

This means that any manipulation of the env state that happens on the env object will be considered the **latest** version of state before we commit changes back to disk.

Reply:
I'm still a little confused about this.
Could you help me understand what's the use case here?

What I was imagining was something like:

> The environment file is updated outside of `azd` scope. For example, from a hook. The hook might've added new key-values or updated them.  Then, the next time `azd` runs `save()`, are we overriding any changes from azd as the latest?

Quick sample
1. azd runs -> creates env file -> [ foo=bar ]
2. hook runs -> updates env file -> [ foo=bar, hook=xyz ]
3. azd continues -> What's env in memory for azd here? 
  1. [ foo=bar ] ?
  2.  [ foo=bar, hook=xyz ] ?
4. azd continue -> updates env in memory -> call save -> cash -> reload from file -> override with cash

So, do we need to do the `cash->override` in case `azd continues` starts by doing env->reload to pick the env just like it is coming from the hook?

Reply:
After #3 you have 2 instances of env... The one updated in the hook via a `azd env set ...` command and the one from the calling `azd`.  The original `azd` call does not know that the env was updated, but the next call will typically be a new command that went through the hooks middleware and was reloaded.

I do like the idea of just eventually getting rid of the `Save` command all together but that also means we'll need a more Generic `SetValue()` func to replace the direct map mutations like we have today.  I'm hesitant to make that big of a change right now but would for sure put it in for GA.

Reply:
example flow:

`azd deploy`
1. Hooks middleware would run first bootstrapping original `env`
2. `pre` hooks runs (which would reload `env` again
3. deploy command runs (Some service targets will append new vars)
4. deploy saves `env` (resulting in another reload
5. `post` hook runs (which would reload `env` again